### PR TITLE
Code reorg proposal: isolate liquid transformation code

### PR DESCRIPTION
**Pull request by EvidenceBKidscode**
Hello,

This PR is for discussing about code reorganization of liquid transformation.

We had many changes for Kidscode on liquid transformation and the fist thing we did was to isolate all liquid transformation code.

I think this is sane to do that. I'd like to discuss about it. Should the same change be done in Mintetest ?

This PR does not add or remove any functionnality, it is only about code reorganization (and actually works).

It is not yet possible to change the liquid logic, anyway there is only the classic one yet (even if it's very easy to implement a LiquidLogicNone that makes liquids still), but it makes a big step toward that possibility.

Here is a summary of what is proposed.

**Two new classes**
- ```LiquidLogic```, a virtual class for any liquid logic;
- ```LiquidLogicClassic```, a class for actual Minetest liquid logic.

**```ReflowScan``` removed**
```ReflowScan``` has been entirely removed. Its code is now in ```LiquidLogicClassic``` class:
- ```ReflowScan::scan``` --> ```LiquidLogicClassic::scanBlock```
- ```ReflowScan::lookupBlock``` --> ```LiquidLogicClassic::lookupBlock```
- ```ReflowScan::isLiquidFlowableTo``` --> ```LiquidLogicClassic::isLiquidFlowableTo```
- ```ReflowScan::isLiquidHorizontallyFlowable``` --> ```LiquidLogicClassic::isLiquidHorizontallyFlowable```

**Changes on ```Map``` class**
- New member ```m_liquid_logic``` added. It represents the liquid logic used for this map.
- All code and members of ```Map``` class for liquid transformations has been moved to ```LiquidLogicClassic```
- ```Map::transformLiquids``` method has been moved to ```LiquidLogicClassic::transform```

**Changes on ```Mapgen``` class**
```Mapgen``` class has been updated to use map liquid logic. Some of its methods moved to ```LiquidLogicClassic```:
- ```Mapgen::updateLiquid``` --> ```LiquidLogicClassic::scanVoxelManip```
- ```Mapgen::isLiquidHorizontallyFlowable``` --> ```LiquidLogicClassic::isLiquidHorizontallyFlowable```

### CODE REVIEW

**Comment by SmallJoker** [comment]
I'm happy to see `map.cpp` shrink in size. It makes sense to bundle everything liquid flow related into one class. This PR is only about moving code, so I'll :+1: after checking that it still works ingame.

**Comment by nerzhul** [comment]
It a good idea to refactor the map class which is quite heavy. If you can rebase it can be nice. Also i don't like the "logic" name, please call it Mgr like other of our managers"

File: src/liquidlogic.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "liquidlogic.h"
+#include "map.h"
+#include "nodedef.h"
+
+LiquidLogic::LiquidLogic(Map *map, IGameDef *gamedef) :
+	m_map(map),
+	m_gamedef(gamedef)
+{
+	m_ndef = m_map->getNodeDefManager();
```
Comment (nerzhul) [inline src/liquidlogic.cpp:28]: just put it properly with upper lines using the map pointer instead of m_map

**Comment by pyrollo** [comment]
Sorry for late answer.
Rebasing this PR implies some work. It's not just adapting the code to new version, it's kind of redoing the whole code reorg.
There's no problem getting into it, but I'd like to be sure it is interesting core devs. (IMO it is important to separate liquid management from map).

**Comment by sfan5** [comment]
Is there some specific reason to organize the class hierarchy so that liquid logic could possibly be swapped out? I mean I read the description, but do you have an usecase where you want to swap out the whole liquid logic?

Other than that moving all liquid stuff to a single file is a good idea and I support it.

**Comment by pyrollo** [comment]
This code reorg is a pre-condition for implementing  other liquid managments.

In Kidscode we developed a different liquid management, a true finite liquid management. I guess this finite liquid management could be interesting for some Minetest games (actually does not work well with current mapgens, oceans are leaking into caves).

I did some tests, having both finite and classic liquid managements, depending on a node def field, but it was too bad on a performance point of view. It could be a per-world setting.

**Comment by rubenwardy** [comment]
Closed due to inactivity

**Pull request closed without merging**